### PR TITLE
Add new flag for whitelisting node taints

### LINKED ISSUE #74282: Support custom taints during e2e tests
**What would you like to be added**:
Support for users with clusters that have `NoSchedule` taints on specific nodes other than master nodes so that they can still run e2e tests.

**Why is this needed**:
It is a reoccurring problem that Sonobuoy users report problems with e2e tests failing to run on their clusters because of a setup which included custom taints.

 - Having looked briefly into the code it seems like plumbing a custom, allowed NoSchedule taint would be pretty extensive, is that correct?
 - Is the idea of _only_ supporting the master NoSchedule taint something that is deliberate/documented somewhere that I have not found it?
 - Is there a recommended workaround? If we assume their taints on the nodes are not flexible, the only option they would have would be to configure the test context to set `TestContext.AllowedNotReadyNodes=(number of non-master, unschedable nodes)`, is that correct? Is there a reason they shouldn't try and do that?

/sig testing

### PR DESCRIPTION
**Pull request by johnSchnake**
**What type of PR is this?**
/kind feature

Technically a feature I suppose since it adds a new flag; considered a bug in the test framework by some.

**What this PR does / why we need it**:
Adds a new flag which allows users to specify a regexp
which will effectively whitelist certain taints and
allow the test framework to startup tests despite having
tainted nodes.

Fixes an issue where e2e tests were unable to be run
in tainted environments due to the framework waiting for
all the nodes to be schedulable without any tolerations.

**Which issue(s) this PR fixes**:
Fixes #74282

**Special notes for your reviewer**:
 - Moves a few pieces of logic into the `e2enode`. In some cases the logic was duplicated there and in the `framework` package but I just exported it from the `e2enode` package in order to remove the redundancy.
 - Wrapped the main logic change into a single function which was easily testable; added a table driven test for it

**Does this PR introduce a user-facing change?**:
```release-note
/test/e2e/framework: Adds a flag "non-blocking-taints" which allows tests to run in environments with tainted nodes. String value should be a comma-separated list.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:
```docs
NONE
```

### CODE REVIEW

File: test/e2e/framework/node/resource.go
Comment (timothysc) [inline test/e2e/framework/node/resource.go]: I'm worried if this will work on all the pod-specs generated.  

What I'm wondering is if we whitelist we apply a blanket toleration.  

If a test tries to schedule a pod won't it just hang...?

Comment (johnSchnake) [reply test/e2e/framework/node/resource.go]: I think there are two separate use cases with different solutions:
 - a user has a cluster with N nodes, M of which are tainted with their own custom NoSchedule taints (e.g. node/etcd, node/control-plane, etc) which are meant to be NoSchedule. They do not want test pods to be allowed to run on those nodes, they just need the existing "wait until ALL nodes are scheduleable unless they have the master noe label" logic to tolerate them.
 - a user has a cluster with N nodes, all N are tainted with various NoSchedule taints but in their actual use, they deliberately taint pods they wish to run on certain nodes.

This PR (and I thought the issue) was around solving the first issue; that is the one that I've seen and heard from users about.

You've mentioned case 2, but I'm not sure that's the right problem to solve here:
 - it would have a potential impact of every test
 - it would require an explicit list of exact taints which I've been told may be frustrating to provide whereas this currently allows a regexp (this was a comment in the DD)
 - I am admittedly unfamiliar with this use case, but it seems unclear that it would work uniformly for all those users. What if they have N nodes, M tainted for reason X, (N-M) tainted for reason Y (gpu availability, networking capabilities, geography, etc). Would they want the test suite to run any/all workloads on any/all nodes? Maybe the answer is yes, maybe my situation is even a bit silly, but it seems like if they are tainting all their nodes they may need extra strict control over what workloads run where and it is hard to see how/why that should be supported. Especially with the idea of conformance=workload portability; is everyone clear on the expectation that those clusters to be conformant even if we couldn't move a vanilla hello-world pod into them without added tolerations? 

I'm sure you've thought about that more than I have, but that just wasn't the problem I thought I was trying to solve by this ticket.

Comment (timothysc) [reply test/e2e/framework/node/resource.go]: So my thinking was both user stories are valid, and have gotten reports on both.    

The logic for determining nodes ready needs to change but also pod tolerations need to be adjusted.

Comment (johnSchnake) [reply test/e2e/framework/node/resource.go]: If the second use case applies to you, would it be reasonable to require users to add their own mutating webhook before trying to run tests? Isn't that what we'll have to do to pass tests in that situation? In addition, if tests get cut off in the middle, wouldn't it be possible to have left a mutating webhook impacting all pods in the user cluster? That seems like a concern to me.

Regardless, do you have a problem with me continuing with solving use case 1 with a regexp as done?

This would mean that startup isnt blocked by taints matching XYZ, and then to solve use case 2 we'd have to have a separate, concrete list of taints which MUST be tolerated by pods in order for tests to pass.

We don't want to try and reuse a list of taints for both use cases because, as I mentioned above, it seems very possible to have every pod tainted in a custom way but only intend pods to be scheduled on a subset of them.

Comment (timothysc) [reply test/e2e/framework/node/resource.go]: > If the second use case applies to you, would it be reasonable to require users to add their own mutating webhook before trying to run tests? 

It's a pain and it's why we wanted this change.  I'll prod folks from the wild to comment on the issue.

File: test/e2e/framework/node/wait.go
Comment (timothysc) [inline test/e2e/framework/node/wait.go]: What happens with the master tainted nodes?

File: test/e2e/framework/node/wait_test.go
Comment (timothysc) [inline test/e2e/framework/node/wait_test.go]: could you add a test with multiple nodes where one is tainted as master.

**Review by timothysc — CHANGES_REQUESTED** [review]

File: test/e2e/framework/node/resource.go
Comment (oomichi) [inline test/e2e/framework/node/resource.go]: IsNodeUntaintedWhitelist() is called in package node only, so it is not necessary to expose it.

File: test/e2e/framework/node/wait_test.go
Comment (johnSchnake) [reply test/e2e/framework/node/wait_test.go]: This particular method is for testing single nodes; I will refactor the main logic so that this multi-node logic is also testable and add some more test cases.

File: test/e2e/framework/node/wait.go
Comment (johnSchnake) [reply test/e2e/framework/node/wait.go]: I'm not sure I understand the question.

If you have a node that is tainted as `node-role.kubernetes.io/master`, that is also the default for the whitelisted taint regexp. This would mean that we would use the path above (not `else` path this comment is on).

The end result would be that we would just check that it is ready and doesn't have other NoSchedule taints on it besides that one. Since it doesnt, this method would return `true` and the node would not be considered a blocker for the tests starting up.

This will become more clear when I extract the multi-node logic into a more testable situation. That way we can see at a high level when the cluster would start up vs keep polling.

File: pkg/controller/nodeipam/ipam/BUILD
Comment (johnSchnake) [inline pkg/controller/nodeipam/ipam/BUILD]: ugh...Rebase went awry. I think a bunch of wonky things got into the build now (I was just trying to push a 1 word spelling fix)

File: test/e2e/framework/node/resource.go
Comment (johnSchnake) [reply test/e2e/framework/node/resource.go]: Done.

**Comment by johnSchnake** [comment]
/remove sig/api-machinery
/remove sig/apps
/remove sig/cluster-lifecycle 
/remove sig/instrumentation
/remove sig/network

Removing tags placed due to bad rebase (ended up touching a bunch of files I didn't intend to and don't now).

**Comment by johnSchnake** [comment]
@timothysc I believe I've added the requested testing and I've fixed up the build/spelling issues that were blocking the build. Ready for review again.

**Review by timothysc — APPROVED** [review]
/lgtm 
/approve

**Comment by johnSchnake** [comment]
Was in the merge pool but ended up having failures/merge issues. I've rebased onto master and am retesting.

**Comment by johnSchnake** [comment]
A constant value we referenced from another package had been unexported, hence the typecheck failure. It was just used for a single test so I just defined it locally in the `wait_test.go` file.

**Comment by johnSchnake** [comment]
Rebasing _again_ because it says there is a build conflict again. In addition, one of the logger tests are getting failures even though I didnt touch it. Seems like a very odd type of failure to be flaky on so I'm not sure. Maybe the 2 are related and rebasing again will fix the test failure.

**Comment by johnSchnake** [comment]
Small unit test failures have been resolved but there is a more troubling problem with the conformance-kind-ipv6 job. That is timing out while waiting for the nodes to be ready which is directly related to this PR. Investigating.

/test pull-kubernetes-e2e-gce

**Comment by johnSchnake** [comment]
Logs show
```
I0829 17:21:44.735] Aug 29 16:59:14.717: INFO: Unschedulable nodes:
I0829 17:21:44.735] Aug 29 16:59:14.717: INFO: -> kind-control-plane Ready=true Network=false Taints=[{node-role.kubernetes.io/master  NoSchedule <nil>}]
```
Which, to me, indicate that this may have failed because of network issues and not because of the taint. The logs are not clear which of the failure criteria was hit to mark it as unschedulable. Rerunning to see.

/test pull-kubernetes-conformance-kind-ipv6

**Comment by johnSchnake** [comment]
I think ^ may be incorrect and the `Network=false` is really implying `NetworkUnavailable=false` which is desirable. I guess we'll see.

**Comment by johnSchnake** [comment]
I am having problems building locally so I'm pushing a debug commit to get more info about the failure. :(

**Comment by johnSchnake** [comment]
I had some of the logic in a `setupSuite` method which I _thought_ was the right place to put it. However I guess in some paths it doesn't get called. I moved it into a more idiomatic location that is only for updating the testContext after all the flags are parsed. Some tests failing for, what seem to be, unrelated issues. Retesting.

/test pull-kubernetes-conformance-kind-ipv6

File: test/e2e/framework/test_context.go
Comment (andrewsykim) [inline test/e2e/framework/test_context.go]: I think "Whitelist" taints is confusing naming. Instinctively I read it as "Taints that specifically aren't ignored".  A few other names that come to mind:
* non-blocking-taints
* ignored-taints

I'm also wondering why regex? I think most use-cases will be for specific taints like the control plane ones, we should keep it simple and have this a comma delimited list of taint key?

Comment (johnSchnake) [reply test/e2e/framework/test_context.go]: Regex was just because in the early incarnation of this issue I created a DD and the only comment I got was that maybe it would be more versatile if we did a regexp in case people have `foo/x` `foo/y` etc. The indication I got was that there may be some people with a large number of taints, though I'm not personally aware of situations with large # of unique taints.

If we had a static list we could simplify some of the logic and add only a single new field to the testContext instead of the string/regexp pair.

I'm fine with a rename on the variable, its not a big issue to me.`NonBlockingTaints` sgtm

Comment (johnSchnake) [reply test/e2e/framework/test_context.go]: https://github.com/kubernetes/kubernetes/issues/74282#issuecomment-470217210

@neolit123 Can you elaborate on the need? I don't have a hat in the ring on that point (regexp vs list) but you said it wouldn't be ideal to have to explicitly list them. Do you standby that enough that you'd be against the list implementation? Both @timothysc and @andrewsykim think the list is more intuitive/simple/sufficient.

Comment (neolit123) [reply test/e2e/framework/test_context.go]: regexp seems more flexible to me.
in terms of naming of the flag, best to exclude it's "type" of the name e.g. `my-flag-int vs my-flag`

Comment (andrewsykim) [reply test/e2e/framework/test_context.go]: IMO we should stick to list of taints and **if** the use-case comes up, we can adapt it to handle catch-alls for known prefixes as opposed to regex, for example:
```
--non-blocking-taints="node-role.kubernetes.io/master,node-role.kubernetes.io/special-node,my-cloud.io/*"
```

**Comment by xmudrii** [comment]
@andrewsykim @neolit123 This PR fixes the issue that is in the 1.16 milestone, but as the code freeze started, do we want to move both the PR and the issue for 1.17 or this is urgent?

**Comment by neolit123** [comment]
we should move them to 1.17.
/milestone v1.17

File: test/e2e/framework/test_context.go
Comment (johnSchnake) [reply test/e2e/framework/test_context.go]: I feel like @andrewsykim has the more compelling approach here:
 - avoids premature generalization
 - simplifies interface presented to the user (which would also simplify things for the implementation)

@neolit123, you're right its more flexible (which is why I originally wrote it that way based on your input) but I feel like in order to get this to move forward this simpler approach may be more appropriate. Do you disagree [enough to want to '/hold' this if we don't use the regexp approach]?

This has now been moved to v1.17 so it will have to stall for a bit, but I'll make the tweak to just be a string and get ready for the un-freeze. Hopefully I can get review/consensus by then and it will go in right away

Comment (neolit123) [reply test/e2e/framework/test_context.go]: @johnSchnake no strong opinions on my side.
let's defer to the `/test` approvers on the decision here: @timothysc @andrewsykim @oomichi 

> This has now been moved to v1.17 so it will have to stall for a bit

at least, until the 10th of September when code freeze is lifted.

**Comment by alejandrox1** [comment]
Are those failures flakes?
/retest

File: test/e2e/framework/test_context.go
Comment (johnSchnake) [reply test/e2e/framework/test_context.go]: 👍 Changes made to make this a csv list of strings that is passed around. Consistent with other csv flags, we pass around the whole list as a single string and let the user parse it as needed.

**Review by timothysc — APPROVED** [review]
/lgtm

**Comment by andrewsykim** [comment]
@johnSchnake release notes should be updated to reflect the new flag name, not sure I would consider this "user facing" though

**Comment by neolit123** [comment]
> @johnSchnake release notes should be updated to reflect the new flag name, not sure I would consider this "user facing" though

or at least prefixed with something like `/test/e2e/framework:...`

**Pull request merged**
